# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [44]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, r2_score

In [16]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

**Check the shape of your data**

In [17]:
spaceship.shape

(8693, 14)

**Check for data types**

In [18]:
spaceship.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

**Check for missing values**

In [19]:
spaceship.isna().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

In [20]:
spaceship.dropna(inplace = True)
spaceship.shape

(6606, 14)

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [21]:
spaceship['Cabin'] = spaceship['Cabin'].apply(lambda x: x[:1])

- Drop PassengerId and Name

In [22]:
spaceship = spaceship.drop(['PassengerId', 'Name'], axis='columns')

- For non-numerical columns, do dummies.

In [23]:
spaceship.dtypes # homeplanet, cyrosleep, cabin, destination and Vip are categorical

HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Transported        bool
dtype: object

In [31]:
for item in ['HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'VIP']:
    dummies = pd.get_dummies(spaceship[item], prefix = item)
    spaceship = pd.concat([spaceship, dummies], axis=1)
    spaceship.drop(item, axis=1, inplace=True)

spaceship

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0     39.0          0.0        0.0           0.0     0.0     0.0        False   
1     24.0        109.0        9.0          25.0   549.0    44.0         True   
2     58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3     33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4     16.0        303.0       70.0         151.0   565.0     2.0         True   
...    ...          ...        ...           ...     ...     ...          ...   
8688  41.0          0.0     6819.0           0.0  1643.0    74.0        False   
8689  18.0          0.0        0.0           0.0     0.0     0.0        False   
8690  26.0          0.0        0.0        1872.0     1.0     0.0         True   
8691  32.0          0.0     1049.0           0.0   353.0  3235.0        False   
8692  44.0        126.0     4688.0           0.0     0.0    12.0         True   

      HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  Cabin_D  \
0                False               True            False  ...    False   
1                 True              False            False  ...    False   
2                False               True            False  ...    False   
3                False               True            False  ...    False   
4                 True              False            False  ...    False   
...                ...                ...              ...  ...      ...   
8688             False               True            False  ...    False   
8689              True              False            False  ...    False   
8690              True              False            False  ...    False   
8691             False               True            False  ...    False   
8692             False               True            False  ...    False   

      Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0       False    False    False    False                    False   
1       False     True    False    False                    False   
2       False    False    False    False                    False   
3       False    False    False    False                    False   
4       False     True    False    False                    False   
...       ...      ...      ...      ...                      ...   
8688    False    False    False    False                     True   
8689    False    False     True    False                    False   
8690    False    False     True    False                    False   
8691     True    False    False    False                     True   
8692     True    False    False    False                    False   

      Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                         False                     True       True     False  
1                         False                     True       True     False  
2                         False                     True      False      True  
3                         False                     True       True     False  
4                         False                     True       True     False  
...                         ...                      ...        ...       ...  
8688                      False                    False      False      True  
8689                       True                    False       True     False  
8690                      False                     True       True     False  
8691                      False                    False       True     False  
8692                      False                     True       True     False  

[6606 rows x 25 columns]

**Perform Train Test Split**

In [33]:
features = spaceship.drop('Transported', axis='columns')
target = spaceship['Transported']

In [34]:
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size = 0.20, random_state=0)

**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [37]:
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)
pred = knn.predict(X_test)

In [49]:
pred.dtype
y_test.dtype

dtype('bool')

- Evaluate your model's performance. Comment it

In [46]:
print(f'Accuracy = {knn.score(X_test, y_test)}')
print(f"Precision = {precision_score(y_test, pred, average='binary')}")
print(f"Recall = {recall_score(y_test, pred, average='binary')}")
#print(f"R2 = {r2_score(y_test, pred)}")

Accuracy = 0.7692889561270801
Precision = 0.7746913580246914
Recall = 0.7594553706505295


In [47]:
print(f"Accuracy improvement = {(knn.score(X_test, y_test) - 0.7655068078668684) / 0.7655068078668684 * 100}")
print(f"Precision improvement = {(precision_score(y_test, pred, average='binary') - 0.7482319660537482) / 0.7482319660537482 * 100}")
print(f"Recall improvement = {(recall_score(y_test, pred, average='binary') - 0.800302571860817) / 0.800302571860817 * 100}")

Accuracy improvement = 0.4940711462450496
Precision improvement = 3.5362552218254844
Recall improvement = -5.103969754253316


By using the categorical variables for the prediction, the model's accuracy is only improved by 0.5%, the precision by 3.5% and the recall is even worsened by 5%. We conduct that the machine learning algorithm could not be improved by this type of feature engineering.